### ISSUE TITLE
Caching batched search API calls only caches last request in batch and not working for some pnp functions.

### ISSUE THREAD

**Issue by DmitriyVdE** [pnp/pnpjs#2333]
### Category
- [ ] Enhancement
- [x] Bug
- [x] Question
- [ ] Documentation gap/issue

### Version

Please specify what version of the library you are using: [3.4.1]

Please specify what version(s) of SharePoint you are targeting: [SharePoint Online]

### Expected / Desired Behavior / Question
Hi!

I'm using batching and caching (5 min) to reduce the amount of calls to SPO and improve loading times.
I've used caching for individual calls successfully and I've also used batching successfully, combining them has caused me quite a bit of frustration and I just can't seem to get it to work.

### Observed Behavior
I've looked at the documentation [here](https://pnp.github.io/pnpjs/concepts/batching-caching/) and tried to implement the functionality as shown, although using a forEach to add calls to the batch. (See 'Steps to Reproduce')

```
// Relevant imports
import { SPFI, spfi, SPFx } from '@pnp/sp';
import '@pnp/sp/webs';
import '@pnp/sp/search';
import { dateAdd } from '@pnp/core';
import { Caching } from '@pnp/queryable';
// tslint:disable-next-line: no-duplicate-imports
import { ISearchQuery, SearchResults, ISearchResult, SearchQueryBuilder } from '@pnp/sp/search';
import { WebPartContext } from '@microsoft/sp-webpart-base';
```

### Steps to Reproduce

The first problem occurs when batching search calls, these work fine the first time (before caching), but only returns the last result from the batch when retrieving items from the cache.

This code essentially retrieves a bunch of properties from different sites to be used by a component.

```
    const sitePages = SOMEARRAYOFPAGEITEMS[]; // Gets filled in a previous call (Also cached and works correctly)

    const sp = spfi().using(SPFx(context));

    const [batchedSP, execute] = sp.batched();

    const cacheKey2 = `UNIQUEANDPREDICTABLECACHEKEY`;

    batchedSP.using(
      Caching({
        store: 'session',
        keyFactory: () => cacheKey2,
        expireFunc: () => dateAdd(new Date(), 'minute', 5),
      }),
    )

    const batchResults = [];

    sitePages.forEach(page => {
      const queryText = `Path=${page.OriginalPath.substring(
        0,
        page.OriginalPath.indexOf('/Lists'),
      )} ContentClass:STS_Site`;

      batchedSP
        .search({
          Querytext: queryText,
          SelectProperties: [
            'Title',
            'ViewCount',
            ...
          ],
          RowLimit: 1,
        } as ISearchQuery)
        .then(r => {
          batchResults.push(r.PrimarySearchResults[0] as any);
        });
    });

    await execute();

    // Do something with batchResults
```

My second issue occurs when I try to retrieve the likes and comments from an array of site pages using a batch with caching from a specific web. In this case the cached 'value' array is simply empty and caching does nothing.

```
// Function receives the webUrl and a list of pages.

    const sp = spfi(webUrl).using(SPFx(this._spContext));
    const [batchedSP, execute] = sp.batched();

    const cacheKey = `ANOTHERUNIQUEANDPREDICTABLECACHEKEY`;

    batchedSP.using(
      Caching({
        store: 'session',
        keyFactory: () => cacheKey,
        expireFunc: () => dateAdd(new Date(), 'minute', 5),
      }),
    )

    const pagesWithLikesAndComments: INews[] = [];
    const activities: any[] = [];

    // * Add a batch call for the LikesCount and the CommentsCount with respective callbacks to add it to the results array
    pages.forEach(page=> {
      // * Likes
      batchedSP.web.lists
        .getById(page.ListId)
        .items.getById(page.ListItemId)
        .getLikedBy()
        .then(r => {
            // Do something with the result -> push to pagesWithLikesAndComments & activities
        });

      // * Comments
      batchedSP.web.lists
        .getById(page.ListId)
        .items.getById(page.ListItemId)
        .comments.expand('replies', 'likedBy', 'replies/likedBy')()
        .then(r => {
            // Do something with the result -> push to pagesWithLikesAndComments & activities
        });
    });

    // * Executes the batched calls
    await execute();

    // Do something with pagesWithLikesAndComments & activities
```

So finally my question is, am I doing something wrong, or are these simply things that aren't implemented correctly/at all?
I could write some caching myself but since the function is available in @pnp/sp I'd like to use it.
If I missed anything or anyone has some further questions about my implementation please feel free to ask.

I'd like to thank anyone answering in advance!

**Comment by juliemturner** [comment]
I don't think I understand what this means...
>but only returns the last result from the batch when retrieving items from the cache.

But I think you should first abandon the forEach and do normal for loops and see if that doesn't fix the issue.

```ts
pages.forEach(page=> {
   ...
}
```

```ts
for(let i=0; i<pages.length; i++){
   const page = pages[i];
   ....
}
```

**Comment by DmitriyVdE** [comment]
So I replaced the forEach loops with normal for loops and it didn't change the behavior. Below an example of the empty 'value' for the batched '.getLikedBy()' and '.comments.expand('replies', 'likedBy', 'replies/likedBy')()' calls.

![image](https://user-images.githubusercontent.com/43698501/176257487-cb991acf-60ca-4c8b-87aa-918c0d487ebc.png)
![image](https://user-images.githubusercontent.com/43698501/176255769-f88b1592-a992-4935-af4f-3b0a08ec8c18.png)

This is really weird since a value is being returned by the function. (3 runs with different cacheKeys)

![image](https://user-images.githubusercontent.com/43698501/176257283-ec0ad76a-52c9-4c34-b82d-b96f2ec3f0a6.png)


Then for the batched search calls, check the screenshots below to see what I mean by 'only the last result gets returned'.

Initial call result:
![image](https://user-images.githubusercontent.com/43698501/176256873-6b9ecdd7-9ac9-4289-9412-c47540c64bc1.png)

Cached result:
![image](https://user-images.githubusercontent.com/43698501/176256646-b739c94e-58da-4af1-a67e-c3deb20d13fb.png)

As you can see, the array in the cached result is just filled with the last item from the result of the initial call.

**Comment by juliemturner** [comment]
Ok, thanks for doing the extra work, we'll look into it and see if we can figure out if there's a bug or not.

**Comment by patrick-rodgers** [comment]
Hi @DmitriyVdE,

Couple of things happening here, one is a bug in the library, and the other is something I think we can make more clear in the documentation. Sorry for your frustration.

**Search Caching**
There was a bug in the search caching logic where the same key was always being used. This is because 1) we create dynamic keys for every request from the url, and 2) the url for all search queries are the same so the key always ended up the same. Meaning no matter how many search requests you did the last one "won" and overwrote whatever was already cached, and then that result was always returned based on the non-unique key for ALL search requests.

Got this fixed and it will be in 3.5 coming soon (likely next week), you can see the updates in PR #2344.

**Caching + Batching**
Here I believe the documentation can be clearer. When you specify the `cacheKey` and `cacheKey2` values, those are used for all the requests. That key function is meant to be determined per request, even for batched requests we cache on a per-request basis. I'll update the docs to clarify.

This is our default cache key implementation, note how we use the URL of the request to make a key. You would want your function to be unique per request, or just let us handle it.

```TS
keyFactory: (url: string) => getHashCode(url.toLowerCase()).toString()
```

If you leave off the keyFactory parameter in your code, does it then start to cache correctly? In my tests it looks like everything is working, my testing code is below:

```TS
// I didn't test in SPFx, your setup is fine.
const sp = spfi("{URL}").using(SPBrowser());

const [batchedSP, execute] = sp.batched();

// I passed no options, you can pass any of the options or none
batchedSP.using(Caching());

const items = await sp.web.lists.getByTitle("Generic").items.top(20)();

const batchedList = batchedSP.web.lists.getByTitle("Generic");

const itemsValues = [];
const fieldValues = [];

for (let i = 0; i < items.length; i++) {

    const item = items[i];

    batchedList.items.getById(item.Id).select("Title", "PicField")().then((ii) => {
        itemsValues.push(ii);
    });

    batchedList.items.getById(item.Id).fieldValuesAsText().then((ii) => {
        fieldValues.push(ii);
    });

}

await execute();
```

**Comment by DmitriyVdE** [comment]
Hi @patrick-rodgers

Thank you for fixing this issue!

I wasn't aware that I didn't have to pass any options to 'Caching()' or even how the keyFactory worked and put a string in there. Thank you for explaining this!

I'll try these fixes when the next release comes out and get back to you after some testing since I have different tasks to complete right now.
